### PR TITLE
:seedling: support clusterctl generate cluster with templates from stdin

### LINKED ISSUE #7227: Support standard input as input source for `clusterctl generate cluster` command
**User Story**

As a developer/user, I would like to pass cluster templates to the `clusterctl generate cluster` command via stdin for the following reasons:
- allow this syntax: `$any_command | clusterctl generate cluster --from -`, so that the generate command can be a part of a whole pipeline of steps
- make it easier to use clusterctl with cluster templates that are in a different machine. e.g. during development, I edit a cluster-template.yaml file locally, then pass it via stdin to `clusterctl` running on a remote host. otherwise, this would be two steps (copy the cluster template file, then run clusterctl generate cluster).

**Detailed Description**

Currently, `clusterctl generate cluster` accepts the following template sources:
- `[-i $provider] [-f $flavor]`: parse templates from released `cluster-template-${flavor}.yaml` files from the configured infrastructure providers.
- `--from $url`: parse template from github, raw URL or a local file
- `--from-config-map $cm`: parse template from a configmap in the management cluster.

I would like to suggest adding support for `clusterctl generate cluster --from -`, which would read the cluster template from stdin instead. Implementation can be quite simple, since all we need to do is handle `-` as a value in https://github.com/kubernetes-sigs/cluster-api/blob/cede911727f3cba3a402348a7c6e636d7a44ab6a/cmd/clusterctl/client/cluster/template.go#L132

**Anything else you would like to add:**

Something like this is already supported in the `clusterctl generate yaml` command, but that one has slightly different semantics from `clusterctl generate cluster` (e.g. no cluster_name, namespace, etc).

/kind feature

### PR DESCRIPTION
**Pull request by neoaggelos**
**What this PR does / why we need it**:

Support stdin as an input source for the `clusterctl generate cluster` command. See #7227 for reference.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7227

### CODE REVIEW

**Review by ykakarap — COMMENTED** [review]
The overall approach seems good. Can you also please add some test cases.

**Comment by neoaggelos** [comment]
@ykakarap added a simple test case, let me know if more are required.

File: cmd/clusterctl/client/cluster/template.go [hunk 1]
```diff
@@ -129,6 +130,14 @@ func (t *templateClient) GetFromURL(templateURL, targetNamespace string, skipTem
 }
 
 func (t *templateClient) getURLContent(templateURL string) ([]byte, error) {
+	if templateURL == "-" {
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/template.go:133]: what about documenting this option in the command help and into https://cluster-api.sigs.k8s.io/clusterctl/commands/generate-cluster.html#alternative-source-for-cluster-templates

**Review by fabriziopandini — COMMENTED** [review]
nice work, only one nit from my side

File: cmd/clusterctl/client/cluster/template.go (hunk 1 above)
Comment (neoaggelos) [reply cmd/clusterctl/client/cluster/template.go:133]: Thanks, I've updated the docs and added a small note in the help message of the flag.

File: cmd/clusterctl/client/cluster/template_test.go [hunk 2]
```diff
@@ -307,6 +307,10 @@ func Test_templateClient_GetFromURL(t *testing.T) {
 	path := filepath.Join(tmpDir, "cluster-template.yaml")
 	g.Expect(os.WriteFile(path, []byte(template), 0600)).To(Succeed())
 
+	// redirect stdin
+	os.Stdin, err = os.Open(path)
```
Comment (ykakarap) [inline cmd/clusterctl/client/cluster/template_test.go:311]: Please adda  defer block to reset the value of `os.Stdin`.

**Review by ykakarap — COMMENTED** [review]
Just one comment. LGTM otherwise.

**Comment by fabriziopandini** [comment]
lgtm pending addressing last nit and squash

File: cmd/clusterctl/client/cluster/template_test.go (hunk 2 above)
Comment (neoaggelos) [reply cmd/clusterctl/client/cluster/template_test.go:313]: Did it in a very simplistic way, hope it's enough.

**Review by ykakarap — APPROVED** [review]
/lgtm

File: cmd/clusterctl/client/cluster/template_test.go (hunk 2 above)
Comment (ykakarap) [reply cmd/clusterctl/client/cluster/template_test.go:313]: I think it is enough.

**Comment by fabriziopandini** [comment]
@neoaggelos thanks for your first contribution in Cluster API!
/lgtm
/approve

**Pull request merged**